[P105] Move AHT10/20/21 plugin from Playground

tonhuisman opened the pull request:
Move plugin for AHT10/20/21 I2C environment sensor from the Playground.

- Some adjustments to the code, reformatted using Uncrustify
- Re-used Plugin ID P105
- Documentation

TODO:
- [x] Complete documentation
- [ ] Testing (to be reported)
  - [x] AHT10 (confirmed by ghtester, and dragonfly69)
  - [ ] AHT20
  - [x] AHT21

Resolves #3604 
Resolves #3466
Resolves #3436 
Resolves #3550

tonhuisman committed:
[P105] Move AHT10 plugin from PlayGround

TD-er commented:
Ah, glad it was already reviewed on the Playground :)

Make sure to add big warnings for the "can only be used as the only sensor on the I2C bus"

tonhuisman committed:
[P105] Check for other I2C devices and show warning on device settings

tonhuisman committed:
[P105] Complete the documentation

tonhuisman committed:
[P105] Update Custom-sample.h plugin comment

tonhuisman committed:
[P105] Add AHT10 to I2C scanner

tonhuisman commented:
Hm, I have chosen the ATH10-only implementation, mostly because it was the last one added to the Playground, but looking closer at the implementation that **sakinit** did in _P248_AHT.ino, also supporting the more modern AHT20 and AHT21, it feels appropriate to use _that_ version (though it has been partly replaced in the Playground, Github still has the old files, of course).
I'll work on that tomorrow, and update documentation too.

TD-er commented:
OK, let me know which one you would like to be merged.

tonhuisman committed:
Merge branch 'mega' into feature/p105-move-aht10-from-playground

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/p105-move-aht10-from-playground

tonhuisman committed:
[P105] Replaced code by the 'other' AHT plugin from Playground, updated documentation, some other doc. fixes

tonhuisman commented:
Switched to the _P248_AHT source from Plugin Playground, as that seems a little more structured, and also supports the more modern, and better behaving on the I2C bus, AHT20/AHT21 sensors.
Updated the documentation accordingly.

Also fixed a few minor issues and small typos in newly merged documentation, reported by Sphinx, not enough to validate a separate PR.

TD-er commented on src/_P105_AHT.ino:
```diff
@@ -0,0 +1,206 @@
+      }
+      {
+        const String options[] = { F("AHT10"), F("AHT20"), F("AHT21") };
+        int indices[]          = { AHT10_DEVICE, AHT20_DEVICE, AHT21_DEVICE };
```
Just nitpicking, but these indices can also be defined as `const`.
Maybe the compiler then can optimize it slightly more.

tonhuisman replied:
I'll use that more often then: no need to change -> make `const` 😃

TD-er commented on src/src/PluginStructs/P105_data_struct.h:
```diff
@@ -1,99 +1,81 @@
+  AHT21_DEVICE = 21,
+};
 
+enum AHTx_state {
```
Can you make this `enum class` ? This does add the requirement to prefix the enum values with `AHTx_state::` and may require some more `static_cast` calls, but then the compiler can help detect if you missed handling an element in a switch statement and also gives a compiler error if you provide the wrong type as a function argument.
Otherwise it is just an `int` and you can easily mix up enum values order in a function call.

TD-er commented on src/src/PluginStructs/P105_data_struct.cpp:
```diff
@@ -1,262 +1,191 @@
+  value         = data[3] & 0x0F;
+  value         = (value << 8) | data[4];
+  value         = (value << 8) | data[5];
+  last_temp_val = static_cast<float>(value) / (1 << 20) * 200 - 50;
```
Please use braces here. Probably not needed, but I'm always triggered by these so it slows down reading the source code ;)

TD-er commented on src/src/PluginStructs/P105_data_struct.cpp:
```diff
@@ -1,262 +1,191 @@
-}
+  value        = (value << 8) | data[2];
+  value        = (value << 4) | (data[3] >> 4);
+  last_hum_val = static_cast<float>(value) / (1 << 20) * 100;
```
Same here... do you mean to divide it by roughly 100 million, or divide by a million and multiply by 100 ?

tonhuisman replied:
That's (all) inherited code... only changed the casting from the original Playground code...

Will fix after walking the dog 🐕

tonhuisman replied:
The formulas provided in the data sheet look like this:

![Screenshot - 01_08_2021 , 20_06_29](https://user-images.githubusercontent.com/1922393/127781039-5c5bcbb7-3ade-4446-95a6-02c678287c5a.png)

tonhuisman committed:
[P105] Code improvements, based on feedback

tonhuisman committed:
[P105] Update I2C Scanner to include AHT20/AHT21

TD-er commented on src/src/PluginStructs/P105_data_struct.cpp:
```diff
@@ -30,15 +30,15 @@ AHTx_Device::AHTx_Device(uint8_t addr, AHTx_device_type type) :
+    case AHTx_device_type::AHT10_DEVICE: return F("AHT10");
+    case AHTx_device_type::AHT20_DEVICE: return F("AHT20");
+    case AHTx_device_type::AHT21_DEVICE: return F("AHT21");
     default: return F("AHTx");
```
You could leave out the "default" part and just return this outside the switch scope.
Then the compiler will warn you when you missed one (if you add one later)
Using `default:` removes this kind of checks as the default option makes all possible values being handled.

tonhuisman committed:
[P105] Small code improvement

ghtester commented:
_Just testing with custom build, compiled with this PR. Unfortunately I can't get AHT10 working - it's not even detected by I2C Scan, tested alone and with LCD2004 plugin with the same result. There're no messages in log regarding to AHT10. I'll try to struggle with pull-up resistors but maybe I did not compile it correctly as I am using Custom.h and there're recent changes in this template._

Edit - fixed, it was a bad wire in my test circuit. So I can confirm AHT10 works, quickly tested together with LCD2004 plugin and both seems to be working fine. Thanks for a great work! :-)

TD-er commented:
Given the tricky behavior of the AHT10, you may want to power cycle it, don't perform an I2C scan, but just test it as the only connected I2C device.

ghtester commented:
@TD-er Thanks for an advice, tested several things and in the end found a bad wire junction... So the P105 plugin works fine, at least so far and at least with AHT10 device. :-)

TD-er commented:
Great :)

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/p105-move-aht10-from-playground

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/p105-move-aht10-from-playground

tonhuisman committed:
[P105] Small improvements

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/p105-move-aht10-from-playground

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/p105-move-aht10-from-playground

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/p105-move-aht10-from-playground

ghtester commented:
Unfortunately it can't be compiled anymore:

```
Compiling .pio/build/custom_IR_ESP8266_4M1M/src/ESPEasy.ino.cpp.o
/home/vagrant/GitHub/letscontrolit/ESPEasy/src/ESPEasy.ino:18:5: warning: #warning "**** Using Settings from Custom.h File ***" [-Wcpp]
   # warning "**** Using Settings from Custom.h File ***"
     ^
/home/vagrant/GitHub/letscontrolit/ESPEasy/src/_P105_AHT.ino: In function 'boolean Plugin_105(uint8_t, EventStruct*, String&)':
/home/vagrant/GitHub/letscontrolit/ESPEasy/src/_P105_AHT.ino:90:10: error: 'PLUGIN_I2C_GET_ADDRESSES_HEX' was not declared in this scope
     case PLUGIN_I2C_GET_ADDRESSES_HEX:
          ^
*** [.pio/build/custom_IR_ESP8266_4M1M/src/ESPEasy.ino.cpp.o] Error 1
========================================== [FAILED] Took 268.38 seconds ==========================================

Environment             S
```

tonhuisman commented:
>  it can't be compiled anymore:

I'll fix after committing some other WIP.
It had an early version of the I2CScanner changes, merged last week.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/p105-move-aht10-from-playground

tonhuisman committed:
[P105] Fix I2Cscanner merge issue

ghtester commented:
Compiled... Thanks a lot!

tonhuisman committed:
Merge branch 'mega' into feature/p105-move-aht10-from-playground

tonhuisman commented:
I finally wired the AHT21 I received last week, and connected it. Works like charm so far (though only a short testing time yet).
And as this has actually the same internals as an AHT20, I _think_ that one can also get a ✔️ in the list above. Though anyone that owns such sensor is of course invited to report their test results.

I'd say this one is ready to merge.
